Complete docstring URLs

mmatera opened the pull request:
This just completes with WR links almost all the remaining docstrings.

mmatera commented:
I used the following code to add the field where it was missing. Then, I reverted the spurious changes by hand.


```
import os
from mathics.builtin import modules, Builtin

def ensure_url(module):
    filename = module.__file__
    new_file_lines = []
    cls = None
    docline = ""
    inside_docstr = False
    inside_class = False
    class_indent_level = 0
    changed = False
    print("Processing file ", filename)

    with open(filename, "r") as f:
        for line in f.readlines():
            # docstrings are processed inside
            # class
            
            if inside_docstr:
                docline = docline + line
                try:
                    eval(docline)
                except SyntaxError:
                    continue
                inside_docstr = False
                continue
            # Not the docstring, just add
            # the line
            strip_line = line.strip()
            if inside_class:
                # This tries to process the docline of
                # the class.
                if strip_line != "":
                    if docline == "":
                        docline = strip_line
                        if docline == "":
                            continue
                        if len(docline)>1 and docline[0] in ("f", "r"):
                            docline = docline[1:]
                        if docline[0] in ("'", '"'):
                            try:
                                eval(docline)
                            except SyntaxError:
                                inside_docstr = True
                            continue
                        docline = "''"
                    # Check if we still are inside the class by checking
                    # the indent level of the line
                    curr_indent = len(line) - len(line.lstrip())
                    if curr_indent < class_indent_level:
                        inside_class = False
                        docline = ""
                        class_indent_level = 0
                new_file_lines.append(line)
                continue
                        
            if strip_line.startswith("class "):
                class_indent_level = len(line) - len(line.lstrip()) + 4
                new_file_lines.append(line)
                name = strip_line[6:].split("(")[0]
                name = name.split(":")[0]
                try:
                    cls = module.__dict__[name]
                except:
                    print(f"<<{name}>>", "is not the name of a class.")
                    raise
                doc_str = cls.__doc__
                if doc_str is not None:
                    indent_spaces = (class_indent_level * " ")
                    if issubclass(cls, Builtin):
                        if doc_str.count("<url>:") == 0:
                            print("missing url in", cls)
                            changed = True
                            url_str = f"{indent_spaces}<url>:WMA link:https://reference.wolfram.com/language/ref/{name}.html</url>\n"
                            doc_str = url_str + doc_str
                            
                    if "\\" in doc_str:
                        doc_str = doc_str.replace("\\", "\\\\")
                    doc_str = f'{indent_spaces}"""\n{doc_str}"""\n'
                    try:
                        eval(doc_str)
                    except SyntaxError:
                        doc_str =  indent_spaces + "r" + doc_str.lstrip()
                    try:
                        eval(doc_str)
                    except SyntaxError:
                        print("<<",doc_str,">>  " , "is not a valid string")
                        raise
                    doc_str = doc_str  
                    new_file_lines.append(doc_str)
                inside_class = True
                docline = ""
                inside_docstr = False
                continue
            new_file_lines.append(line)
    if changed:
        print("------------Changed!------------")
        rewritten_file = filename+"-new"
        with open(rewritten_file, "w") as f:
            for line in new_file_lines:
                f.write(line)
        os.rename(filename, filename + "~")
        os.rename(rewritten_file, filename)        
```

rocky commented on mathics/builtin/attributes.py:
```diff
@@ -276,14 +282,16 @@ class HoldAllComplete(Predefined):
 
 class HoldFirst(Predefined):
     """
+    <url>:WMA link:https://reference.wolfram.com/language/ref/HoldFirst.html</url>
```
This causes a duplicate link.

rocky commented on mathics/builtin/list/eol.py:
```diff
@@ -475,17 +483,18 @@ def apply(self, expr, evaluation):
     <dl>
       <dt> FirstCase[{$e1$, $e2$, ...}, $pattern$]
-      <dd>gives the first $ei$ to match $pattern$, or $Missing[\"NotFound\"]$ if none matching pattern is found.
+      <dd>gives the first $ei$ to match $pattern$, or $Missing["NotFound"]$ if none matching pattern is found.
```
Mistake from before  `$Missing["NotFound"]$` -> `'Missing["NotFound"]'`

rocky commented on mathics/builtin/list/eol.py:
```diff
@@ -475,17 +483,18 @@ def apply(self, expr, evaluation):
 
-      <dt>FirstCase[$expr$, $pattern$, $default$, $levelspec$] \
-      <dd>finds only objects that appear on levels specified by $levelspec$.
+      <dt>FirstCase[$expr$, $pattern$, $default$, $levelspec$]       <dd>finds only objects that appear on levels specified by $levelspec$.
```
Line join added here.

mmatera replied:
Notice that this is due to a wrong backslash in the master version.

rocky commented on mathics/builtin/numbers/constants.py:
```diff
@@ -195,6 +195,8 @@ def to_sympy(self, expr=None, **kwargs):
 
 class Catalan(_MPMathConstant, _SympyConstant):
     """
+    <url>:WMA link:https://reference.wolfram.com/language/ref/Catalan.html</url>
```
Duplicate line

rocky reviewed (changes requested):
Please stop with the update for more until we've gone over the first batch.

rocky commented:
> I used the following code to add the field where it was missing. Then, I reverted the spurious changes by hand.
> 
> ```
> import os
> from mathics.builtin import modules, Builtin
> 
> def ensure_url(module):
>     filename = module.__file__
>     new_file_lines = []
>     cls = None
>     docline = ""
>     inside_docstr = False
>     inside_class = False
>     class_indent_level = 0
>     changed = False
>     print("Processing file ", filename)
> 
>     with open(filename, "r") as f:
>         for line in f.readlines():
>             # docstrings are processed inside
>             # class
>             
>             if inside_docstr:
>                 docline = docline + line
>                 try:
>                     eval(docline)
>                 except SyntaxError:
>                     continue
>                 inside_docstr = False
>                 continue
>             # Not the docstring, just add
>             # the line
>             strip_line = line.strip()
>             if inside_class:
>                 # This tries to process the docline of
>                 # the class.
>                 if strip_line != "":
>                     if docline == "":
>                         docline = strip_line
>                         if docline == "":
>                             continue
>                         if len(docline)>1 and docline[0] in ("f", "r"):
>                             docline = docline[1:]
>                         if docline[0] in ("'", '"'):
>                             try:
>                                 eval(docline)
>                             except SyntaxError:
>                                 inside_docstr = True
>                             continue
>                         docline = "''"
>                     # Check if we still are inside the class by checking
>                     # the indent level of the line
>                     curr_indent = len(line) - len(line.lstrip())
>                     if curr_indent < class_indent_level:
>                         inside_class = False
>                         docline = ""
>                         class_indent_level = 0
>                 new_file_lines.append(line)
>                 continue
>                         
>             if strip_line.startswith("class "):
>                 class_indent_level = len(line) - len(line.lstrip()) + 4
>                 new_file_lines.append(line)
>                 name = strip_line[6:].split("(")[0]
>                 name = name.split(":")[0]
>                 try:
>                     cls = module.__dict__[name]
>                 except:
>                     print(f"<<{name}>>", "is not the name of a class.")
>                     raise
>                 doc_str = cls.__doc__
>                 if doc_str is not None:
>                     indent_spaces = (class_indent_level * " ")
>                     if issubclass(cls, Builtin):
>                         if doc_str.count("<url>:") == 0:
>                             print("missing url in", cls)
>                             changed = True
>                             url_str = f"{indent_spaces}<url>:WMA link:https://reference.wolfram.com/language/ref/{name}.html</url>\n"
>                             doc_str = url_str + doc_str
>                             
>                     if "\\" in doc_str:
>                         doc_str = doc_str.replace("\\", "\\\\")
>                     doc_str = f'{indent_spaces}"""\n{doc_str}"""\n'
>                     try:
>                         eval(doc_str)
>                     except SyntaxError:
>                         doc_str =  indent_spaces + "r" + doc_str.lstrip()
>                     try:
>                         eval(doc_str)
>                     except SyntaxError:
>                         print("<<",doc_str,">>  " , "is not a valid string")
>                         raise
>                     doc_str = doc_str  
>                     new_file_lines.append(doc_str)
>                 inside_class = True
>                 docline = ""
>                 inside_docstr = False
>                 continue
>             new_file_lines.append(line)
>     if changed:
>         print("------------Changed!------------")
>         rewritten_file = filename+"-new"
>         with open(rewritten_file, "w") as f:
>             for line in new_file_lines:
>                 f.write(line)
>         os.rename(filename, filename + "~")
>         os.rename(rewritten_file, filename)        
> ```

While this is good, I recommend doing in the future is to try this on a small human-reviewable sample, and then go over this carefully. And then put in a PR for that. 

Apparently there are bugs in this code when we the URL line is split as happens for Catalan's constant.

BTW - have you tried building the LaTeX doc to see if there are problems with that?  (There might have been problems from before this, but compounding potentially massively if there is a problem makes fixing LaTeX doc problems worse.)

(Some of this

mmatera commented:
The idea of using code to try to complete something that should be almost mechanical was to find a pattern that then can be also (in a coarse grained way) tested automatically. But OK, let's clean this part, and then the remaining 64 cases.

rocky commented:
> The idea of using code to try to complete something that should be almost mechanical was to find a pattern that then can be also (in a coarse grained way) tested automatically.

That is a fine thing to do. It is the process of doing and the code that is could be improved. Run the program on a small sample, find the mistakes and then enlarge. 

I suppose you must have looked at this and found some bugs and did this. So then the next step is to put in a PR for a relatively small sample so that others might find things that were missed. 

When I have a large collection of things to update I usually go in orders of magnitude. so maybe 10 files first then 100. Then the rest.

mmatera commented:
OK, I think I have removed all the duplicated links, and also restore some changes from raw str to str.  Also, I fixed a couple of mistakes in docstrings.
Pdf documentation (make doc) is properly built too. 

I can split this into smaller bunches, but I guess that seeing all the changes together is easier (since all follow the same pattern).

rocky commented:
> OK, I think I have removed all the duplicated links, and also restore some changes from raw str to str. Also, I fixed a couple of mistakes in docstrings. Pdf documentation (make doc) is properly built too.
> 
> I can split this into smaller bunches, but I guess that seeing all the changes together is easier (since all follow the same pattern).

Thanks - would you split this up into a first batch of 10 files, that we or I can go over in fine detail. Thanks.

rocky commented:
Closing because this is covered by other commits. However you can run the program again and find stragglers. Beware though of false positives that the program find.

mmatera commented:
@rocky, @TiagoCavalcante, thank you very much for the work of revision and correction.  Regarding the program, it worked very well regarding what was written for: propel the job that @rocky asked to do before the release, and unveil some of the implicit patterns that we have in the documentation :)

rocky commented:
> @rocky, @TiagoCavalcante, thank you very much for the work of revision and correction. Regarding the program, it worked very well regarding what was written for: propel the job that @rocky asked to do before the release, and unveil some of the implicit patterns that we have in the documentation :)

Yes, it worked well enough for that. The process could be improved so that it is less chaotic and catastrophic. One misfeature of the program when multiplied by lots of files means a lot of additional work.  

I have a PR in for isort importing. But the difference there is that this isn't some homegrown piece of code that has not been testing a lot -  looking at it stats: https://pepy.tech/project/isort it is used a lot and has something like 5K stars.  That kind of thing I am okay running over the code base.

A program that one of us writes though doesn't start out in that league and so we shouldn't treat it like that.

mmatera commented:
I would be happy to use a standard code to do the job, but for this concrete task, it does not exist. If all the docstrings would have a field of the form <URL>:..:...</URL>  I could have used `sed` to change and reformat them in a more or less standard and automatic way, so my attempt was just to complete this aspect.

rocky commented:
> I would be happy to use a standard code to do the job, but for this concrete task, it does not exist.

That is understood. This is more about adapting as the situation arises. When code is likely to have problems, _whereever_ it comes from, or understanding _exactly_ what needs to be done is open-ended or murky, then tread _lightly_.   Do some thing small, get review, and then go larger.
